messagequeue/messagequeue.go

```diff
@@ -16,63 +21,172 @@ var log = logging.Logger("bitswap")
+	r.current.RemoveType(c, wtype)
+}
+
 // New creats a new MessageQueue.
 func New(ctx context.Context, p peer.ID, network MessageNetwork) *MessageQueue {
-	return &MessageQueue{
+	return newMessageQueue(ctx, p, network, maxMessageSize, sendErrorBackoff)
+}
+
+// This constructor is used by the tests
+func newMessageQueue(ctx context.Context, p peer.ID, network MessageNetwork, maxMsgSize int, sendErrorBackoff time.Duration) *MessageQueue {
+	mq := &MessageQueue{
 		ctx:                 ctx,
-		wl:                  wantlist.NewSessionTrackedWantlist(),
-		network:             network,
 		p:                   p,
+		network:             network,
+		maxMessageSize:      maxMsgSize,
+		bcstWants:           newRecallWantList(),
+		peerWants:           newRecallWantList(),
+		cancels:             cid.NewSet(),
 		outgoingWork:        make(chan struct{}, 1),
 		done:                make(chan struct{}),
 		rebroadcastInterval: defaultRebroadcastInterval,
+		sendErrorBackoff:    sendErrorBackoff,
+		priority:            maxPriority,
 	}
+
+	// Apply debounce to the work ready signal (which triggers sending a message)
+	debounced := debounce.New(sendMessageDebounce)
+	mq.signalWorkReady = func() { debounced(mq.onWorkReady) }
+
+	return mq
 }
 
-// AddMessage adds new entries to an outgoing message for a given session.
-func (mq *MessageQueue) AddMessage(entries []bsmsg.Entry, ses uint64) {
-	if !mq.addEntries(entries, ses) {
+// Add want-haves that are part of a broadcast to all connected peers
+func (mq *MessageQueue) AddBroadcastWantHaves(wantHaves []cid.Cid) {
+	if len(wantHaves) == 0 {
 		return
 	}
-	select {
-	case mq.outgoingWork <- struct{}{}:
-	default:
+
+	mq.wllock.Lock()
+	defer mq.wllock.Unlock()
+
+	for _, c := range wantHaves {
+		mq.bcstWants.Add(c, mq.priority, pb.Message_Wantlist_Have)
+		mq.priority--
 	}
+
+	// Schedule a message send
+	mq.signalWorkReady()
+}
+
+// Add want-haves and want-blocks for the peer for this message queue.
+func (mq *MessageQueue) AddWants(wantBlocks []cid.Cid, wantHaves []cid.Cid) {
+	if len(wantBlocks) == 0 && len(wantHaves) == 0 {
+		return
+	}
+
+	mq.wllock.Lock()
+	defer mq.wllock.Unlock()
+
+	for _, c := range wantHaves {
+		mq.peerWants.Add(c, mq.priority, pb.Message_Wantlist_Have)
+		mq.priority--
+	}
+	for _, c := range wantBlocks {
+		mq.peerWants.Add(c, mq.priority, pb.Message_Wantlist_Block)
+		mq.priority--
+	}
+
+	// Schedule a message send
+	mq.signalWorkReady()
 }
 
-// AddWantlist adds a complete session tracked want list to a message queue
-func (mq *MessageQueue) AddWantlist(initialWants *wantlist.SessionTrackedWantlist) {
-	initialWants.CopyWants(mq.wl)
-	mq.addWantlist()
+// Add cancel messages for the given keys.
+func (mq *MessageQueue) AddCancels(cancelKs []cid.Cid) {
+	if len(cancelKs) == 0 {
+		return
+	}
+
+	mq.wllock.Lock()
+	defer mq.wllock.Unlock()
+
+	for _, c := range cancelKs {
+		mq.bcstWants.Remove(c)
+		mq.peerWants.Remove(c)
+		mq.cancels.Add(c)
```

Comment:
Right - because of rebroadcasting, I think the only thing we can do reliably is not add the cancel if the want has never been sent (it's not in `bcstWants.allWants` and not in `peerWants.allWants`). But that seems unlikely to happen in practice.